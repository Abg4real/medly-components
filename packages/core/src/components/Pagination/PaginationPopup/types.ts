export interface Props {
    prevPageNumber: number;
    nextPageNumber: number;
    onClickHandler: (page: any) => void;
}
