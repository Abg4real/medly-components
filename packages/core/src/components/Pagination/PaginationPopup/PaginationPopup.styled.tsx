import { centerAligned, styled } from '@medly-components/utils';
import { getPageNumberButtonStyleByState, PageNumberButton } from '../Pagination.styled';

export const PaginationBackgroundStyled = styled('div')`
    position: fixed;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background: transparent;
    z-index: 1000;
    ${centerAligned()}
`;

export const PageEllipsisOverlay = styled('div')`
    max-height: 20rem;
    display: flex;
    align-items: center;
    flex-direction: column;
    overflow: auto;
    background: #fff;
    box-shadow: 0 2px 8px #b0bcc8;
    border-radius: 4px;

    & > ${PageNumberButton} {
        display: block;
        line-height: 3.2rem;
        border-radius: 0;
        margin: 0;
        height: 4rem;
        width: 4rem;
        padding: 0.4rem;

        &:hover {
            ${getPageNumberButtonStyleByState('overlayPageNumber', 'hovered')}
        }

        &:active {
            ${getPageNumberButtonStyleByState('overlayPageNumber', 'hovered')}
        }
    }
`;
