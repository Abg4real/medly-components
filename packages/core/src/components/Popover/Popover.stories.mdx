import { Popover, PopoverWrapper } from './Popover';
import { Preview, Story, Meta, Props } from '@storybook/addon-docs/blocks';
import * as stories from './Popover.stories';

<Meta title="Core/Popover" component={Popover} />

# Popover

You have to use both `Popover` & `PopoverWrapper` component.

```typescript
import { Popover, PopoverWrapper } from '@medly-components/core';
```

<Preview withToolbar>
    <Story name="Basic">{stories.Basic()}</Story>
</Preview>

### PopoverWrapper Props

<Props of={PopoverWrapper} />

### Popover Props

<Props of={Popover} />
