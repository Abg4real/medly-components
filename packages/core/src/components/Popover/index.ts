export { default as Popover } from './Popover';
export { default as PopoverWithBackground } from './PopoverWithBackground';
export { default as PopoverWrapper } from './PopoverWrapper';
