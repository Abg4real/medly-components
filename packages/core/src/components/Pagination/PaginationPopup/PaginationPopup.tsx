import React, { FC, useContext } from 'react';
import Popover from '../../Popover';
import { addPageItems } from '../helper';
import { PageNumberButton } from '../Pagination.styled';
import { PageEllipsisOverlay, PaginationBackgroundStyled } from './PaginationPopup.styled';
import { Props } from './types';

export const PaginationPopup: FC<Props> = React.memo(({ prevPageNumber, nextPageNumber, onClickHandler }) => {
    const [isPopoverVisible, setPopoverVisibility] = useContext(Popover.Context),
        onClickHandlerWrapper = (pageNumber: any) => () => {
            setPopoverVisibility(false);
            onClickHandler(pageNumber);
        };
    return (
        <>
            <PageNumberButton key={prevPageNumber + '...'}>...</PageNumberButton>
            {isPopoverVisible && <PaginationBackgroundStyled />}
            <Popover.Popup>
                <PageEllipsisOverlay>
                    {addPageItems(prevPageNumber + 1, nextPageNumber - 1, []).map(item => {
                        return (
                            <PageNumberButton key={item} onClick={onClickHandlerWrapper(item)}>
                                {item}
                            </PageNumberButton>
                        );
                    })}
                </PageEllipsisOverlay>
            </Popover.Popup>
        </>
    );
});
