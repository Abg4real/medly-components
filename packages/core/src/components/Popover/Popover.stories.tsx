import { centerAligned, styled } from '@medly-components/utils';
import { boolean, select } from '@storybook/addon-knobs';
import { useState } from '@storybook/addons';
import React from 'react';
import Popover from './Popover';
import PopoverWrapper from './PopoverWrapper';
import { InteractionType, Placement } from './PopoverWrapper/types';

export const placement: Placement[] = [
    'top-start',
    'top',
    'top-end',
    'right-start',
    'right',
    'right-end',
    'bottom-start',
    'bottom',
    'bottom-end',
    'left-end',
    'left',
    'left-start'
];

const interactionType: InteractionType[] = ['click', 'hover'];

const DummyDiv = styled('div')`
    color: white;
    width: 200px;
    height: 100px;
    background-color: steelblue;
    border: 1px solid grey;
    display: inline-block;
    font-size: 20px;
    border-radius: 5px;
    ${centerAligned()};
`;

const DummyWrapper = styled('div')`
    width: 100%;
    height: 100%;
    ${centerAligned()};
`;

const DummyPopover = styled('div')`
    padding: 10px;
    color: black;
    border-radius: 5px;
    border: 1px solid grey;
    ${centerAligned()};
`;

export const Basic = () => {
    const [showPopover, setPopoverState] = useState(false);
    const changePopoverState = () => setPopoverState(val => !val);

    return (
        <DummyWrapper>
            <PopoverWrapper
                showPopover={showPopover}
                placement={select('Placement', placement, 'bottom')}
                interactionType={select('Interaction Type', interactionType, 'hover')}
            >
                <DummyDiv onClick={changePopoverState}>Hello</DummyDiv>
                <Popover fullWidth={boolean('Popover Full Width', false)} fullHeight={boolean('Popover Full Height', false)}>
                    <DummyPopover>This is Popover</DummyPopover>
                </Popover>
            </PopoverWrapper>
        </DummyWrapper>
    );
};
