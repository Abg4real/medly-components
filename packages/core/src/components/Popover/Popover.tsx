import { useOuterClickNotifier, WithStyle } from '@medly-components/utils';
import React, { SFC, useRef, useState } from 'react';
import { PopoverStyled, PopoverWrapperStyled } from './Popover.styled';
import { PopoverProps, PopoverWrapperProps } from './types';

export const PopoverWrapper: SFC<PopoverWrapperProps> & WithStyle = props => {
    const wrapperRef = useRef(null);
    const [popoverState, setPopoverState] = useState(false);
    const { onOuterClick, ...restProps } = props;

    const changePopoverState = (state: boolean) => {
        if (props.interactionType === 'click') {
            const length = wrapperRef.current.children.length - 1;
            const popover = wrapperRef.current.children[length] as HTMLElement;
            popover.style.display = state ? 'block' : 'none';
            setPopoverState(state);
        }
    };

    const handleOnClick = () => {
        changePopoverState(!popoverState);
    };

    useOuterClickNotifier(() => {
        changePopoverState(false);
        onOuterClick && onOuterClick();
    }, wrapperRef);

    return <PopoverWrapperStyled {...restProps} ref={wrapperRef} onClick={handleOnClick} data-testid="popover-wrapper" />;
};

PopoverWrapper.defaultProps = {
    interactionType: 'hover',
    placement: 'bottom'
};
PopoverWrapper.Style = PopoverWrapperStyled;

export const Popover: SFC<PopoverProps> & WithStyle = React.forwardRef((props, ref) => <PopoverStyled ref={ref} {...props} />);

Popover.Style = PopoverStyled;
Popover.defaultProps = {
    fullWidth: false,
    fullHeight: false
};
