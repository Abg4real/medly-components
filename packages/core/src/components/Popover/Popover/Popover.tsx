import { WithStyle } from '@medly-components/utils';
import React, { SFC } from 'react';
import { PopoverStyled } from './Popover.styled';
import { Props } from './types';

export const Popover: SFC<Props> & WithStyle = React.memo(
    React.forwardRef((props, ref) => <PopoverStyled ref={ref} {...props} data-testid="medly-popover" />)
);
Popover.displayName = 'Popover';
Popover.Style = PopoverStyled;
Popover.defaultProps = {
    fullWidth: false,
    fullHeight: false
};
