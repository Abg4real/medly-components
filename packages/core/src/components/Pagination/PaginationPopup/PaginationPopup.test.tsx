import { cleanup, fireEvent, render } from '@test-utils';
import React from 'react';
import PaginationPopup from '.';
import Popover from '../../Popover';

describe('Pagination popup component', () => {
    const renderer = (props: any = { prevPageNumber: 1, nextPageNumber: 3 }) => {
        const mockOnPageClick = jest.fn();
        return {
            mockOnPageClick,
            ...render(
                <Popover interactionType="click">
                    <PaginationPopup {...props} onClickHandler={mockOnPageClick} />
                </Popover>
            )
        };
    };

    afterEach(cleanup);

    it('should render correctly with all the default props', () => {
        const { container } = renderer();
        expect(container).toMatchSnapshot();
    });

    it('should call onClick handler with expected page when selected from ellipsis popover', () => {
        const { mockOnPageClick, getByText } = renderer();
        fireEvent.click(getByText('...').closest('button'));
        fireEvent.click(getByText('2').closest('button'));
        expect(mockOnPageClick).toBeCalledWith(2);
    });
});
