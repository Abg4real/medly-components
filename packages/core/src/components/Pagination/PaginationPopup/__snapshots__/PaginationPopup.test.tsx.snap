// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination popup component should render correctly with all the default props 1`] = `
.c0 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c1 {
  height: 4rem;
  width: 4rem;
  display: block;
  border-radius: 50%;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border: none;
  padding: 0;
  font-weight: 600;
  font-size: 1.2rem;
  line-height: 2rem;
  text-align: center;
  -webkit-transition: all 100ms ease-out;
  transition: all 100ms ease-out;
  color: #435465;
  background: #ffffff;
  margin: 4px;
  height: 3.2rem;
  width: 3.2rem;
  font-weight: 600;
  color: #435465;
  background: #ffffff;
}

.c1:hover {
  cursor: pointer;
}

.c1:focus {
  outline: none;
}

.c1:disabled {
  cursor: not-allowed;
}

.c1:hover {
  color: #303C48;
  background: #F7F8F9;
}

.c1:active {
  color: #13181D;
  background: #eff2f4;
  font-weight: bold;
}

<div>
  <div
    class="c0"
    id="medly-popover-wrapper"
  >
    <button
      class="sc-AxgMl c1"
    >
      ...
    </button>
  </div>
</div>
`;
